wrappers/golang/curves/bls12377/vec_ops.go

```diff
@@ -53,3 +53,26 @@ func VecOp(a, b, out core.HostOrDeviceSlice, config core.VecOpsConfig, op core.V
 
 	return ret
 }
+
+func TransposeMatrix(in, out core.HostOrDeviceSlice, columnSize, rowSize int, ctx cr.DeviceContext, onDevice, isAsync bool) (ret core.IcicleError) {
+
+	var inPointer, outPointer unsafe.Pointer
+	if onDevice {
+		inPointer = in.(core.DeviceSlice).AsPointer()
+		outPointer = out.(core.DeviceSlice).AsPointer()
+	} else {
+		inPointer = unsafe.Pointer(&in.(core.HostSlice[ScalarField])[0])
+		outPointer = unsafe.Pointer(&out.(core.HostSlice[ScalarField])[0])
+	}
```

Comment:
I would ensure that both in and out are indeed on the device using `IsOnDevice()` interface method and also adding `checkDevice` to ensure it is being used on the same device that the memory is located on. 

Maybe the above logic should be in a TransposeCheck function in the core package?

Reply:
right so I should make a transposeCheck function in core that asserts that onDevice is consistent with IsOnDevice(), performs a checkDevice, and picks the right pointer type.
Do I just return the pointers and assign them to inPointer and outPointer ?

Reply:
>Do I just return the pointers and assign them to inPointer and outPointer ?

For now we just perform the checks in `core.xxxCheck` and access the correct pointer type in the curve's function. @nonam3e has an open [PR](https://github.com/ingonyama-zk/icicle/pull/473) to change how we access the pointer, so we will move to that method once it is merged